qiskit/circuit/library/logical_expression_oracle.py

```diff
@@ -0,0 +1,175 @@
+    `&` (`AND`), `|` (`OR`) and `~` (`NOT`),
+    as well as symbols for literals (variables).
+    For example, `'a & b'`, and `(v0 | ~v1) & (~v2 & v3)`
+    are both valid string representation of boolean logical expressions.
+
+    For convenience, this oracle, in addition to parsing arbitrary logical expressions,
+    also supports input strings in the `DIMACS CNF format
+    <http://www.satcompetition.org/2009/format-benchmarks2009.html>`__,
+    which is the standard format for specifying SATisfiability (SAT) problem instances in
+    `Conjunctive Normal Form (CNF) <https://en.wikipedia.org/wiki/Conjunctive_normal_form>`__,
+    which is a conjunction of one or more clauses, where a clause is a disjunction of one
+    or more literals.
+
+    The following is an example of a CNF expressed in DIMACS format:
+
+    .. code:: text
+
+      c This is an example DIMACS CNF file with 3 satisfying assignments: 1 -2 3, -1 -2 -3, 1 2 -3.
+      p cnf 3 5
+      -1 -2 -3 0
+      1 -2 3 0
+      1 2 -3 0
+      1 -2 -3 0
+      -1 2 3 0
+
+    The first line, following the `c` character, is a comment. The second line specifies that the
+    CNF is over three boolean variables --- let us call them  :math:`x_1, x_2, x_3`, and contains
+    five clauses.  The five clauses, listed afterwards, are implicitly joined by the logical `AND`
+    operator, :math:`\land`, while the variables in each clause, represented by their indices,
+    are implicitly disjoined by the logical `OR` operator, :math:`lor`. The :math:`-` symbol
+    preceding a boolean variable index corresponds to the logical `NOT` operator, :math:`lnot`.
+    Character `0` (zero) marks the end of each clause.  Essentially, the code above corresponds
+    to the following CNF:
+
+    :math:`(\lnot x_1 \lor \lnot x_2 \lor \lnot x_3)
+    \land (x_1 \lor \lnot x_2 \lor x_3)
+    \land (x_1 \lor x_2 \lor \lnot x_3)
+    \land (x_1 \lor \lnot x_2 \lor \lnot x_3)
+    \land (\lnot x_1 \lor x_2 \lor x_3)`.
+
+    This is an example showing how to search for a satisfying assignment to an SAT problem encoded
+    in DIMACS using the `Logical Expression oracle with the Grover algorithm.
+    <https://github.com/Qiskit/qiskit-tutorials-community/blob/master/optimization/grover.ipynb>`__
+
+    Logic expressions, regardless of the input formats, are parsed and stored as a source string,
+    from which the corresponding circuits are constructed by
+    :class:`~qiskit.circuit.classicalfunction.ClassicalFunction`. The oracle circuits
+    can then be used with any oracle-oriented algorithms when appropriate. For example, an oracle
+    built from a DIMACS input can be used with the Grover's algorithm to search for a satisfying
+    assignment to the encoded SAT instance.
+    """
+
+    def __init__(self, expression: str) -> None:
+        """
+        Args:
+            expression: The logical expression string.
+        """
+        from qiskit.circuit.classicalfunction import ClassicalFunction
```

Comment:
I think this import can be outside of the class, or does it give circular imports?

Reply:
Ya, it gives circular imports if I put this outside of the class.